flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/stream/MaintenanceTaskBuilder.java

```diff
@@ -0,0 +1,238 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.flink.maintenance.stream;
+
+import java.time.Duration;
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.streaming.api.datastream.DataStream;
+import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
+import org.apache.iceberg.flink.TableLoader;
+import org.apache.iceberg.flink.maintenance.operator.TaskResult;
+import org.apache.iceberg.flink.maintenance.operator.Trigger;
+import org.apache.iceberg.flink.maintenance.operator.TriggerEvaluator;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+
+public abstract class MaintenanceTaskBuilder<T extends MaintenanceTaskBuilder> {
```

Comment:
Do we want to allow the users to create their own maintenance tasks?

Reply:
Let's keep this package private first so that it is easier to evolve the class.

when there is real need in the future, we can always make it public then.

Reply:
Let's mark it experimental then. I definitely would like to provide a way for the users to extend the maintenance tasks.

Reply:
This API is a single `abstract DataStream<TaskResult> append(DataStream<Trigger> sourceStream);` method.
I don't expect big changes here.

Reply:
The public methods are public API anyways...
These are the ones which will be called by the users when they are scheduling the MaintenanceTasks...

Reply:
It would be nice to have this API public in some form. We do have a use case for a custom maintenance-like task (retention, in our case), and it would be really great to be able to integrate with the existing maintenance API instead of making something custom.